File: src/viola_integration.py
"""Integration with Viola CLI for mod merging."""
import os
import shutil
import subprocess
from typing import Callable, List, Optional


def shlex_quote(s):
    """Quote a string for shell usage."""
    try:
        import shlex
        return shlex.quote(str(s))
    except Exception:
        return str(s)


class ViolaIntegration:
    """Handles Viola CLI operations for mod merging."""
    
    def __init__(self, log_callback: Optional[Callable[[str], None]] = None):
        """
        Initialize Viola integration.
        
        Args:
            log_callback: Optional callback function for logging messages
        """
        self.log_callback = log_callback or (lambda x: None)
        self._running = False
        self._proc = None
    
    def merge_mods(self, violacli_path: str, cfgbin_path: str, 
                   mod_paths: List[str], output_dir: str) -> bool:
        """
        Merge mods using Viola CLI.
        
        Args:
            violacli_path: Path to Viola.CLI-Portable.exe
            cfgbin_path: Path to cpk_list.cfg.bin
            mod_paths: List of mod directory paths to merge
            output_dir: Output directory for merged files
            
        Returns:
            True if merge succeeded, False otherwise
        """
        if not os.path.exists(violacli_path):
            self._log("Error: violacli.exe not found")
            return False
        
        if not os.path.exists(cfgbin_path):
            self._log("Error: cpk_list.cfg.bin not found")
            return False
        
        output_dir = os.path.abspath(output_dir)
        os.makedirs(output_dir, exist_ok=True)
        
        cmd = [violacli_path, "-m", "merge", "-p", "PC", "--cl", cfgbin_path] + mod_paths + ["-o", output_dir]
        
        self._log(f"Executing command:\n{' '.join(shlex_quote(x) for x in cmd)}")
        
        try:
            CREATE_NO_WINDOW = getattr(subprocess, "CREATE_NO_WINDOW", 0)
            startupinfo = None
            
            if os.name == "nt":
                try:
                    si = subprocess.STARTUPINFO()
                    si.dwFlags |= subprocess.STARTF_USESHOWWINDOW
                    startupinfo = si
                except Exception:
                    startupinfo = None
            
            self._running = True
            proc = subprocess.Popen(
                cmd,
                stdout=subprocess.PIPE,
                stderr=subprocess.STDOUT,
                text=True,
                creationflags=CREATE_NO_WINDOW if os.name == "nt" else 0,
                startupinfo=startupinfo
            )
            
            self._proc = proc
            
            for line in proc.stdout:
                self._log(line.rstrip())
            
            proc.wait()
            rc = proc.returncode
            self._log(f"violacli finished with code {rc}")
            
            self._proc = None
            self._running = False
            
            return rc == 0
            
        except FileNotFoundError as e:
            self._log(f"Execution error: {e}")
            self._running = False
            return False
        except Exception as e:
            self._log(f"Unexpected error: {e}")
            self._running = False
            return False
    
    def copy_merged_files(self, tmp_data_dir: str, game_data_dir: str) -> bool:
        """
        Copy merged files from temporary directory to game directory.
        
        Args:
            tmp_data_dir: Temporary data directory with merged files
            game_data_dir: Game's data directory
            
        Returns:
            True if copy succeeded, False otherwise
        """
        if not os.path.exists(tmp_data_dir) or not os.path.isdir(tmp_data_dir):
            self._log(f"{tmp_data_dir} was not found. Aborting.")
            return False
        
        self._log(f"Copying {tmp_data_dir} -> {game_data_dir} (overwriting if needed)...")
        
        os.makedirs(game_data_dir, exist_ok=True)
        
        try:
            shutil.copytree(tmp_data_dir, game_data_dir, dirs_exist_ok=True)
        except TypeError:
            # Fallback for older Python versions
            for root, dirs, files in os.walk(tmp_data_dir):
                rel = os.path.relpath(root, tmp_data_dir)
                target_dir = os.path.join(game_data_dir, rel) if rel != "." else game_data_dir
                os.makedirs(target_dir, exist_ok=True)
                for f in files:
                    srcf = os.path.join(root, f)
                    dstf = os.path.join(target_dir, f)
                    try:
                        shutil.copy2(srcf, dstf)
                    except Exception as e:
                        self._log(f"Error copying {srcf} -> {dstf}: {e}")
        
        self._log("Copy completed.")
        return True
    
    def cleanup_temp(self, tmp_data_dir: str) -> bool:
        """
        Clean up temporary directory.
        
        Args:
            tmp_data_dir: Temporary data directory to remove
            
        Returns:
            True if cleanup succeeded, False otherwise
        """
        try:
            shutil.rmtree(tmp_data_dir)
            self._log(f"Removed temporary folder {tmp_data_dir}.")
            return True
        except Exception as e:
            self._log(f"Could not remove {tmp_data_dir}: {e}")
            return False
    
    def is_running(self) -> bool:
        """Check if a merge operation is currently running."""
        return self._running
    
    def stop(self):
        """Stop the current operation if running."""
        if self._proc:
            try:
                self._proc.terminate()
            except Exception:
                pass
            self._proc = None
        self._running = False
    
    def _log(self, message: str):
        """Log a message using the callback."""
        self.log_callback(message)

File: Managers/ViolaIntegration.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace IEVRModManager.Managers
{
    /// <summary>
    /// Provides integration with the Viola CLI tool for merging mods.
    /// </summary>
    public class ViolaIntegration
    {
        private readonly Action<string> _logCallback;
        private readonly Action<int, string>? _progressCallback;
        private Process? _currentProcess;
        private bool _isRunning;

        /// <summary>
        /// Initializes a new instance of the <see cref="ViolaIntegration"/> class.
        /// </summary>
        /// <param name="logCallback">Optional callback for log messages.</param>
        /// <param name="progressCallback">Optional callback for progress updates (progress percentage, status message).</param>
        public ViolaIntegration(Action<string>? logCallback = null, Action<int, string>? progressCallback = null)
        {
            _logCallback = logCallback ?? (_ => { });
            _progressCallback = progressCallback;
        }

        /// <summary>
        /// Merges multiple mods using the Viola CLI tool.
        /// </summary>
        /// <param name="violaCliPath">The path to the violacli.exe executable.</param>
        /// <param name="cfgBinPath">The path to the cpk_list.cfg.bin configuration file.</param>
        /// <param name="modPaths">The list of mod directory paths to merge.</param>
        /// <param name="outputDir">The output directory where merged files will be placed.</param>
        /// <returns><c>true</c> if the merge operation completed successfully; otherwise, <c>false</c>.</returns>
        public async Task<bool> MergeModsAsync(string violaCliPath, string cfgBinPath,
            List<string> modPaths, string outputDir)
        {
            if (!ValidateInputs(violaCliPath, cfgBinPath))
            {
                return false;
            }

            outputDir = Path.GetFullPath(outputDir);
            Directory.CreateDirectory(outputDir);

            var commandArgs = BuildMergeCommandArgs(cfgBinPath, modPaths, outputDir);
            LogCommandExecution(violaCliPath, commandArgs);

            return await ExecuteMergeProcessAsync(violaCliPath, commandArgs);
        }

        private bool ValidateInputs(string violaCliPath, string cfgBinPath)
        {
            if (!File.Exists(violaCliPath))
            {
                _logCallback("Error: violacli.exe not found");
                _progressCallback?.Invoke(0, "ViolaExeNotFound");
                return false;
            }

            if (!File.Exists(cfgBinPath))
            {
                _logCallback("Error: cpk_list.cfg.bin not found");
                _progressCallback?.Invoke(0, "CpkListNotFound");
                return false;
            }

            return true;
        }

        private List<string> BuildMergeCommandArgs(string cfgBinPath, List<string> modPaths, string outputDir)
        {
            var cmd = new List<string> { "-m", "merge", "-p", "PC", "--cl", cfgBinPath };
            cmd.AddRange(modPaths);
            cmd.AddRange(new[] { "-o", outputDir });
            return cmd;
        }

        private void LogCommandExecution(string violaCliPath, List<string> commandArgs)
        {
            var commandLine = string.Join(" ", commandArgs.Select(QuoteArgument));
            _logCallback($"Executing command:\n{violaCliPath} {commandLine}");
            _progressCallback?.Invoke(10, "MergingMods");
        }

        private async Task<bool> ExecuteMergeProcessAsync(string violaCliPath, List<string> commandArgs)
        {
            try
            {
                _isRunning = true;
                var startInfo = CreateProcessStartInfo(violaCliPath, commandArgs);

                _currentProcess = Process.Start(startInfo);
                if (_currentProcess == null)
                {
                    _logCallback("Failed to start violacli process");
                    _progressCallback?.Invoke(0, "FailedToStartViolaProcess");
                    _isRunning = false;
                    return false;
                }

                StartOutputReading();
                var exitCode = await WaitForProcessCompletionAsync();

                return HandleProcessCompletion(exitCode);
            }
            catch (FileNotFoundException ex)
            {
                HandleProcessError($"Execution error: {ex.Message}", $"ExecutionError:{ex.Message}");
                return false;
            }
            catch (Exception ex)
            {
                HandleProcessError($"Unexpected error: {ex.Message}", $"UnexpectedError:{ex.Message}");
                return false;
            }
        }

        private ProcessStartInfo CreateProcessStartInfo(string violaCliPath, List<string> commandArgs)
        {
            return new ProcessStartInfo
            {
                FileName = violaCliPath,
                Arguments = string.Join(" ", commandArgs.Select(QuoteArgument)),
                UseShellExecute = false,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
                CreateNoWindow = true,
                WindowStyle = ProcessWindowStyle.Hidden
            };
        }

        private void StartOutputReading()
        {
            _ = Task.Run(() =>
            {
                if (_currentProcess?.StandardOutput != null)
                {
                    string? line;
                    while ((line = _currentProcess.StandardOutput.ReadLine()) != null)
                    {
                        _logCallback(line);
                    }
                }
            });
        }

        private async Task<int> WaitForProcessCompletionAsync()
        {
            await _currentProcess!.WaitForExitAsync();
            var exitCode = _currentProcess.ExitCode;
            _logCallback($"violacli finished with code {exitCode}");
            return exitCode;
        }

        private bool HandleProcessCompletion(int exitCode)
        {
            _currentProcess = null;
            _isRunning = false;

            if (exitCode == 0)
            {
                _progressCallback?.Invoke(50, "MergeCompletedCopyingFiles");
            }
            else
            {
                _progressCallback?.Invoke(0, $"MergeFailedWithCode:{exitCode}");
            }

            return exitCode == 0;
        }

        private void HandleProcessError(string logMessage, string progressMessage)
        {
            _logCallback(logMessage);
            _progressCallback?.Invoke(0, progressMessage);
            _isRunning = false;
        }

        /// <summary>
        /// Copies merged files from the temporary directory to the game data directory.
        /// </summary>
        /// <param name="tmpDataDir">The temporary directory containing merged files.</param>
        /// <param name="gameDataDir">The game data directory where files will be copied.</param>
        /// <returns><c>true</c> if the copy operation completed successfully; otherwise, <c>false</c>.</returns>
        public bool CopyMergedFiles(string tmpDataDir, string gameDataDir)
        {
            if (!Directory.Exists(tmpDataDir))
            {
                _logCallback($"{tmpDataDir} was not found. Aborting.");
                _progressCallback?.Invoke(0, $"TmpDataNotFound:{tmpDataDir}");
                return false;
            }

            _logCallback($"Copying {tmpDataDir} -> {gameDataDir} (overwriting if needed)...");
            _progressCallback?.Invoke(50, "CopyingFiles");

            Directory.CreateDirectory(gameDataDir);

            try
            {
                var totalFiles = CountFiles(tmpDataDir);
                var copiedFiles = 0;
                CopyDirectory(tmpDataDir, gameDataDir, true, totalFiles, ref copiedFiles);
                _logCallback("Copy completed.");
                _progressCallback?.Invoke(90, "CleaningUpTemporaryFiles");
                return true;
            }
            catch (Exception ex)
            {
                _logCallback($"Error copying files: {ex.Message}");
                _progressCallback?.Invoke(0, $"FailedToCopyMergedFiles:{ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Cleans up the temporary directory by deleting its contents and recreating it.
        /// </summary>
        /// <param name="tmpDir">The temporary directory to clean up.</param>
        /// <returns><c>true</c> if the cleanup completed successfully; otherwise, <c>false</c>.</returns>
        public bool CleanupTemp(string tmpDir)
        {
            try
            {
                if (Directory.Exists(tmpDir))
                {
                    Directory.Delete(tmpDir, true);
                    Directory.CreateDirectory(tmpDir);
                    _logCallback($"Cleaned temporary folder {tmpDir}.");
                }
                return true;
            }
            catch (Exception ex)
            {
                _logCallback($"Could not remove {tmpDir}: {ex.Message}");
                return false;
            }
        }

        /// <summary>
        /// Gets a value indicating whether a merge operation is currently running.
        /// </summary>
        public bool IsRunning => _isRunning;

        /// <summary>
        /// Stops the current merge operation if one is running.
        /// </summary>
        public void Stop()
        {
            if (_currentProcess != null && !_currentProcess.HasExited)
            {
                try
                {
                    _currentProcess.Kill();
                }
                catch
                {
                    // Ignore errors when stopping process
                }
                _currentProcess = null;
            }
            _isRunning = false;
        }

        private static string QuoteArgument(string arg)
        {
            if (string.IsNullOrEmpty(arg))
                return "\"\"";

            if (arg.Contains(' ') || arg.Contains('\t') || arg.Contains('"'))
            {
                return "\"" + arg.Replace("\"", "\\\"") + "\"";
            }

            return arg;
        }

        private void CopyDirectory(string sourceDir, string destDir, bool overwrite, int totalFiles, ref int copiedFiles)
        {
            var dir = new DirectoryInfo(sourceDir);
            Directory.CreateDirectory(destDir);

            CopyFilesInDirectory(dir, destDir, overwrite, totalFiles, ref copiedFiles);
            CopySubdirectories(dir, destDir, overwrite, totalFiles, ref copiedFiles);
        }

        private void CopyFilesInDirectory(DirectoryInfo dir, string destDir, bool overwrite, int totalFiles, ref int copiedFiles)
        {
            foreach (var file in dir.GetFiles())
            {
                var targetFilePath = Path.Combine(destDir, file.Name);
                file.CopyTo(targetFilePath, overwrite);
                copiedFiles++;
                
                UpdateCopyProgress(copiedFiles, totalFiles);
            }
        }

        private void CopySubdirectories(DirectoryInfo dir, string destDir, bool overwrite, int totalFiles, ref int copiedFiles)
        {
            foreach (var subDir in dir.GetDirectories())
            {
                var newDestDir = Path.Combine(destDir, subDir.Name);
                CopyDirectory(subDir.FullName, newDestDir, overwrite, totalFiles, ref copiedFiles);
            }
        }

        private void UpdateCopyProgress(int copiedFiles, int totalFiles)
        {
            // Update progress: 50% to 90% for copying (40% range)
            if (totalFiles > 0 && _progressCallback != null)
            {
                const int copyStartProgress = 50;
                const int copyProgressRange = 40;
                var progress = copyStartProgress + (int)((copiedFiles / (double)totalFiles) * copyProgressRange);
                _progressCallback(progress, $"CopyingFiles:{copiedFiles}:{totalFiles}");
            }
        }

        private static int CountFiles(string directory)
        {
            try
            {
                return Directory.GetFiles(directory, "*", SearchOption.AllDirectories).Length;
            }
            catch
            {
                return 0;
            }
        }
    }
}

